Fix: logical error in grace hash join

devcrafter opened the pull request:
Fixes https://github.com/ClickHouse/ClickHouse/issues/51703

Root cause: number of buckets could become inconsistent if exception was thrown during new buckets creation

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

devcrafter committed:
Try to fix logical error #51703

devcrafter committed:
Number of bucket always increased by 2, so there is no reason provide it as parameter

devcrafter committed:
Merge branch 'master' into fix_logical_error_grace_hash_join

zhanglistar commented:
Check this, already fixed. https://github.com/ClickHouse/ClickHouse/pull/51382

lgbo-ustc commented:
This exception is expected.  There is a certain probalitiy memory allocate failured under stress tests. This make the buckets rehashing fail, and remain the buckets vector in a invalid size.

lgbo-ustc commented:
@devcrafter I think rewirte `rehashBuckets` may be better. Add the new buckets into `GraceHashJoin::buckets` only when all of them are created.

devcrafter committed:
Remove debug tracing

devcrafter committed:
Fix: unexpected number of buckets

Number of buckets could become inconsistent
if exception was thrown during new buckets creation

devcrafter commented:
> @devcrafter I think rewirte `rehashBuckets` may be better. Add the new buckets into `GraceHashJoin::buckets` only when all of them are created.
> 
> And we will see the memory allocate exception instead of the invalid bucket size exception. it's more clear.

Thanks for suggestion. Yes, it's a much more clear approach. Done in [30be0ab](https://github.com/ClickHouse/ClickHouse/pull/51737/commits/30be0ab4a8af7247a12c94076cd17834f712b9d9)

vdimir commented on src/Interpreters/GraceHashJoin.cpp:
```diff
@@ -356,52 +353,66 @@ bool GraceHashJoin::hasMemoryOverflow(const InMemoryJoinPtr & hash_join_) const
-    {
-        auto & left_file = tmp_data->createStream(left_sample_block);
-        auto & right_file = tmp_data->createStream(prepareRightBlock(right_sample_block));
+    // Exception can be thrown in number of cases:
```
If an exception is thrown, won't the query be cancelled? Do we see an inconsistent state of `buckets` because of another thread accessing it when the exception has already been thrown (and the mutex released), but before the query was cancelled?

devcrafter replied:
Yes, that's what I see if intentionally add exception in `addBuckets()` while creating new buckets

devcrafter replied:
Expanded part of `02273_full_sort_join.sql.j2` for grace join:
https://pastila.nl/?00b6056d/cde7c624db96e692a69ef05cc0132f95

Traces for grace hash join with an intentionally added exception when trying to create 5th bucket:
https://pastila.nl/?022538cf/c96e4cb05f99476123270f2ccad3fcdf

Intentionally added exception in log:
```
... <Error> GraceHashJoin: Can't create bucket 4 due to error: Code: 290. DB::Exception: Too many buckets. (LIMIT_EXCEEDED) (version 23.7.1.1)
2
````

lgbo-ustc replied:
The exception is catch and thrown in `PipelineExecutor` asynchronously. I think before the exception is really throw out, other threads still have chance to run and trigger this exception.

devcrafter committed:
Fix style check

devcrafter committed:
Merge remote-tracking branch 'origin/master' into fix_logical_error_grace_hash_join

devcrafter committed:
Merge branch 'master' into fix_logical_error_grace_hash_join

vdimir commented on src/Interpreters/GraceHashJoin.cpp:
```diff
@@ -356,52 +353,66 @@ bool GraceHashJoin::hasMemoryOverflow(const InMemoryJoinPtr & hash_join_) const
+        }
+        catch (...)
+        {
+            LOG_ERROR(
```
Do we need this log? The exception raised by `throw;` below should reach the user and we will see it.

devcrafter replied:
I guess a couple of points:
(1) The exception will not contain the context of the error in an explicit way (can't create bucket), only implicitly as part of callstack.
(2) Without this logging, the log will contain only one exception. But with it, the log will contain all of them, which provide additional information that it happened in different threads.
